omnichain/staking/contracts/Staking.sol

```diff
@@ -29,74 +31,132 @@ contract Staking is ERC20, zContract {
         chainID = chainID_;
     }
 
+    modifier onlySystem() {
+        require(
+            msg.sender == address(systemContract),
+            "Only system contract can call this function"
+        );
+        _;
+    }
+
+    function bytesToBech32Bytes(
+        bytes calldata data,
+        uint256 offset
+    ) public pure returns (bytes memory) {
+        bytes memory bech32Bytes = new bytes(42);
+        for (uint i = 0; i < 42; i++) {
+            bech32Bytes[i] = data[i + offset];
+        }
+
+        return bech32Bytes;
+    }
+
     function onCrossChainCall(
         zContext calldata context,
         address zrc20,
         uint256 amount,
         bytes calldata message
-    ) external override {
+    ) external override onlySystem {
         if (msg.sender != address(systemContract)) {
             revert SenderNotSystemContract();
         }
 
-        address acceptedZRC20 = systemContract.gasCoinZRC20ByChainId(chainID);
-        if (zrc20 != acceptedZRC20) revert WrongChain();
+        if (chainID != context.chainID) {
+            revert WrongChain();
+        }
 
         address staker = BytesHelperLib.bytesToAddress(context.origin, 0);
-        address beneficiary = abi.decode(message, (address));
 
-        stakeZRC(staker, beneficiary, amount);
+        uint8 action = chainID == BITCOIN
+            ? uint8(message[0])
+            : abi.decode(message, (uint8));
+
+        if (action == 1) {
+            stakeZRC(staker, amount);
+        } else if (action == 2) {
+            unstakeZRC(staker);
+        } else if (action == 3) {
+            setBeneficiary(staker, message);
+        } else if (action == 4) {
+            setWithdraw(staker, message, context.origin);
+        } else {
+            revert UnknownAction();
+        }
     }
 
-    function stakeZRC(
-        address staker,
-        address beneficiary,
-        uint256 amount
-    ) internal {
+    function stakeZRC(address staker, uint256 amount) internal {
         stakes[staker] += amount;
-        if (beneficiaries[staker] == address(0)) {
-            beneficiaries[staker] = beneficiary;
-        }
-        lastStakeTime[staker] = block.timestamp;
+        require(stakes[staker] >= amount, "Overflow detected");
```

Comment:
change it to custom errors like the others

Reply:
3a466b3